### ISSUE TITLE
Running code on single GPU

### ISSUE THREAD

**Issue by MLRadfys** [linhuixiao/CLIP-VG#9]
Hi and thanks for you're great work and for making the code public!

I tried to run fully supervised training on a single GPU, but unfortunately the validation values are very low and both training and validation loss are not changing (about the same value).
I started training using the following command:

`python -m torch.distributed.launch --nproc_per_node=1 --master_port 28887 --use_env train_clip_vg.py --num_workers 4 --epochs 110 --batch_size 64 --lr 0.00025  --lr_scheduler cosine --aug_crop --aug_scale --aug_translate    --imsize 224 --max_query_len 77  --sup_type full --dataset unc+     --data_root /path_to_image_data --split_root /path_to_split --output_dir /path_to_output/unc+;`

Now Iam wondering if Iam doing something wrong?

Thanks in advance for any advice,

best regards,

M

**Comment by linhuixiao** [comment]
@MLRadfys  Hi, based on my observation, you may only need to use CUDA_VISIBLE_DEVICES=0 python -m torch.distribute.launch --nproc_per_node=1 --master_port........... for both training and inference on a single GPU.

**Comment by MLRadfys** [comment]
Hi again!

I tested running the code with the above command. Unfortunately Iam still having the same issues.
It seems like the loss and all other metrices are not changing at all. 

Is there anything else I could have missed during installation? I basically just ran `pip install -r requirements.txt`.

I ran the evaluation script using the fully supervised unc+ checkpoint and reached 71% accuracy on the validation dataset. That seems to work :-)

Thanks again,

regards,

M

**Comment by MLRadfys** [comment]
Hi again!

Interestingly, when I reduce the learning rate from 0.00025 to 0.00005 it works.
Does this have something to do with the distributed training?

Thanks again,

kind regards,

M

**Comment by linhuixiao** [comment]
@MLRadfys Hi, I'm glad to see your response. If you can run the inference script, there is a good chance that the training program will work. I tested that a single GPU card works fine in my environment. Regarding your question about reducing the learning rate, may I ask if you have enabled gradient updates for CLIP in the clip_vg.py model file? By default, CLIP should be frozen. Enabling CLIP parameter updates may lead to catastrophic forgetting or performance degradation.

**Comment by MLRadfys** [comment]
@linhuixiao 
Hi and thanks again for you're reply!
I haven't performed any changes in the clip_vg.py file. 
Only the input argument for the learning rate (lr) was changed from 0.00025 to 0.00005. I now managed to perform a fully training of the VIT16 model using the unc+ dataset. It reaches about 73% accuracy on the validation set, which is in line with the reported results.

You mentioned that the CLIP parameters are frozen. Is this done here? 

`for parameter in self.clip.parameters():
            parameter.requires_grad_(False)`

From what I can see, the learning rate for both the CLIP text and vision encoders are set to a very low learning rate. Why is this even needed? 

`param_list = [{"params": rest_param, "lr": args.lr},
                  {"params": visu_cnn_param, "lr": args.lr_visu_cnn},
                  {"params": visu_tra_param, "lr": args.lr_visu_tra},
                  {"params": text_tra_param, "lr": args.lr_bert}]`

 Maybe I miss something here though? :-)

Kind regards,

M
